psl/parser-database/src/attributes/shard_key.rs

```diff
@@ -0,0 +1,167 @@
+use std::borrow::Cow;
+
+use diagnostics::DatamodelError;
+use schema_ast::ast::{self, WithName, WithSpan};
+
+use crate::{
+    attributes::{resolve_field_array_without_args, FieldResolutionError},
+    types::{ModelAttributes, ScalarField, ShardKeyAttribute},
+    Context, ScalarFieldId,
+};
+
+/// `@@shardKey` on models
+pub(super) fn model(model_data: &mut ModelAttributes, model_id: crate::ModelId, ctx: &mut Context<'_>) {
+    let attr = ctx.current_attribute();
+    let fields = match ctx.visit_default_arg("fields") {
+        Ok(value) => value,
+        Err(err) => return ctx.push_error(err),
+    };
+
+    let resolved_fields = match resolve_field_array_without_args(fields, attr.span, model_id, ctx) {
+        Ok(fields) => fields,
+        Err(FieldResolutionError::AlreadyDealtWith) => return,
+        Err(FieldResolutionError::ProblematicFields {
+            unknown_fields: unresolvable_fields,
+            relation_fields,
+        }) => {
+            if !unresolvable_fields.is_empty() {
+                let fields_str = unresolvable_fields
+                    .into_iter()
+                    .map(|((file_id, top_id), field_name)| match top_id {
+                        ast::TopId::CompositeType(ctid) => {
+                            let ct_name = ctx.asts[(file_id, ctid)].name();
+                            Cow::from(format!("{field_name} in type {ct_name}"))
+                        }
+                        ast::TopId::Model(_) => Cow::from(field_name),
+                        _ => unreachable!(),
+                    })
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                let msg = format!("The multi field shard key declaration refers to the unknown fields {fields_str}.");
+                let error =
+                    DatamodelError::new_model_validation_error(&msg, "model", ctx.asts[model_id].name(), fields.span());
+
+                ctx.push_error(error);
+            }
+
+            if !relation_fields.is_empty() {
+                let field_names = relation_fields
+                    .iter()
+                    .map(|(f, _)| f.name())
+                    .collect::<Vec<_>>()
+                    .join(", ");
```

Comment:
we've got itertools here
```suggestion
                    .format(", ")
```